### PR TITLE
Only implement the control protocol

### PR DESCRIPTION
**Pull request by bassosimone**
A follow up pull request will take care of implementing also
other tests like `meta` and `s2c`.

Tested using `measurement_kit ndt` as client.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/ndt-server-go/29)

### CODE REVIEW

File: control/control.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	defer cc.Close()
+
+	rdwr := bufio.NewReadWriter(bufio.NewReader(cc), bufio.NewWriter(cc))
+
+	// Read extended login message
+	login, err := protocol.ReadLogin(rdwr.Reader)
+	if err != nil || login.IsExtended == false {
+		log.Println("ndt: cannot read extended login message")
+		return
+	}
```
Comment (bassosimone) [inline control/control.go]: TODO(bassosimone): make these two distinct error cases

File: control/control.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+		log.Println("ndt: cannot write kickoff message")
+		return
+	}
+	err = rdwr.Writer.Flush()
+	if err != nil {
+		log.Println("ndt: cannot flush kickoff message")
+		return
+	}
+
+	// Queue management (simplified)
```
Comment (bassosimone) [inline control/control.go]: TODO(bassosimone): explain the rationale of this queue management

File: control/control.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	// Send MSG_RESULTS to the client
+	/*
+	 * TODO: Here we should actually send results but to do that we need
+	 * first to implement reading Web100 variables from /proc/web100.
+	 *
+	 * Until we reach this point, send back a variable that NDT client
+	 * will ignore but that is consistent with what it would expect.
+	 */
+	err = protocol.SendSimpleMsg(rdwr.Writer, protocol.MsgResults,
+		"botticelli_does_not_yet_collect_web100_data_sorry: 1\n")
```
Comment (bassosimone) [inline control/control.go]: TODO(bassosimone): do not mention botticelli here

File: control/control.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+// HandleControlConnection handles the control connection |cc|.
+func HandleControlConnection(cc net.Conn) {
+	cc = netx.NewDeadlineConn(cc)
+	defer cc.Close()
+
+	rdwr := bufio.NewReadWriter(bufio.NewReader(cc), bufio.NewWriter(cc))
+
+	// Read extended login message
+	login, err := protocol.ReadLogin(rdwr.Reader)
```
Comment (bassosimone) [inline control/control.go]: TODO(bassosimone): rename `Read` -> `Recv`

File: protocol/protocol.go [hunk 5]
```diff
@@ -78,7 +81,7 @@ const (
 	// SrvQueueServerFault indicates that the session must be terminated.
 	SrvQueueServerFault = "9977"
 	// SrvQueueServerBusy indicates that the server is busy.
-	SrvQueueServerBusy = "9987"
+	SrvQueueServerBusy = "9988"
```
Comment (bassosimone) [inline protocol/protocol.go:85]: TODO(bassosimone): double check this

File: protocol/protocol.go [hunk 6]
```diff
@@ -147,21 +156,28 @@ func ReadLogin(brdr *bufio.Reader) (Login, error) {
 		// TODO(bassosimone): Handle legacy, without json
 		return Login{}, errors.New("not implemented")
 
-	case MsgExtendedLogin:
-		// Handle extended, with json
-		lj := loginJSON{"foo", "bar"}
+	case MsgExtendedLogin: // Handle extended login, i.e. with JSON
+		lj := loginJSON{"foo", "bar"} // Invalid values to catch errors
```
Comment (bassosimone) [inline protocol/protocol.go]: TODO(bassosimone): wouldn't be simpler to leave them empty?

File: protocol/protocol.go [hunk 7]
```diff
@@ -173,6 +189,30 @@ type SimpleMsg struct {
 	Msg string `json:"msg, string"`
 }
 
+const maxMessageLen int = 0xffff
+
+func ReadMessageJson(brdr *bufio.Reader) (Message, error) {
```
Comment (bassosimone) [inline protocol/protocol.go]: TODO(bassosimone): add documentation here

File: protocol/protocol.go [hunk 8]
```diff
@@ -173,6 +189,30 @@ type SimpleMsg struct {
...
+	}
+	simple := &SimpleMsg{}
+	err = json.Unmarshal(msg.Content, &simple)
+	if err != nil {
+		return Message{}, err
+	}
+	nmsg := Message{}
+	nmsg.Header.MsgType = msg.Header.MsgType
+	if len(simple.Msg) > maxMessageLen {
+		panic("unexpected maximum message length")
```
Comment (bassosimone) [inline protocol/protocol.go]: TODO(bassosimone): explain why this is a panic

File: protocol/protocol.go [hunk 9]
```diff
@@ -173,6 +189,30 @@ type SimpleMsg struct {
...
+	}
+	nmsg.Header.Length = int16(len(simple.Msg))
+	// TODO(bassosimone): understand whether this is a string copy and
+	// if we can avoid this copy by using other types.
```
Comment (bassosimone) [inline protocol/protocol.go]: TODO(bassosimone): investigate this

File: protocol/protocol.go [hunk 10]
```diff
@@ -203,5 +243,14 @@ func SendJSON(wr *bufio.Writer, t byte, msg interface{}) error {
 		log.Println(err)
 		return err
 	}
+	if len(j) > maxMessageLen {
+		return errors.New("message is too long")
+	}
 	return Send(wr, t, j)
 }
+
+// TODO(bassosimone): shouldn't we use recv/send or read/write?
```
Comment (bassosimone) [inline protocol/protocol.go]: Short answer: yes

**Review by bassosimone — COMMENTED** [review]
I have added some comments indicating what I need to do to make the PR ready. 👨‍🏭

File: protocol/protocol.go (hunk 5 above)
Comment (bassosimone) [reply protocol/protocol.go:85]: is correct

**Review by bassosimone — COMMENTED** [review]
second pass

**Comment by gfr10598** [comment]
Review status: 0 of 1 LGTMs obtained (waiting on @bassosimone)

---

*[control/control.go, line 27 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/29#-LEFSDBbM8z5iCr2MC54-r1-27:-LEqDb_Z9IhuCpaNtUc9:bg1azdn) ([raw file](https://github.com/m-lab/ndt-server-go/blob/a92591bbe0b7b4e9bb357317b7aef3aeffb9d0be/control/control.go#L27)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

TODO(bassosimone): rename `Read` -> `Recv`
</blockquote></details>

Ah - I see that these are TODOs you left for yourself, that are now resolved.  Can you resolve them or do I have to?

---

*[control/control.go, line 117 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/29#-LEqENUc8zXiy4yRmi9N:-LEqENUc8zXiy4yRmi9O:bi3mgsm) ([raw file](https://github.com/m-lab/ndt-server-go/blob/37e6a43eb54e1d7209858173d64afcd38f185c2b/control/control.go#L117)):*
> ```Go
> 	//
> 	// TODO(bassosimone): Here we should actually send results but to do that we
> 	// need first to implement reading Web100 variables from /proc/web100.
> ```

This is no longer correct.  Want to update to refer to tcp_info?

---

*[protocol/protocol.go, line 207 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/29#-LEFSDBbM8z5iCr2MC52-r1-207:-LEqG0Nj49PXNkMe4ey2:bffbydb) ([raw file](https://github.com/m-lab/ndt-server-go/blob/a92591bbe0b7b4e9bb357317b7aef3aeffb9d0be/protocol/protocol.go#L207)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

TODO(bassosimone): explain why this is a panic
</blockquote></details>

I don't think you should be panicing.  Let's discuss.

---

*[protocol/protocol.go, line 211 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/29#-LEFSDBbM8z5iCr2MC52-r1-211:-LEqGAjsDlvAnchRIMZX:b-58avq4) ([raw file](https://github.com/m-lab/ndt-server-go/blob/a92591bbe0b7b4e9bb357317b7aef3aeffb9d0be/protocol/protocol.go#L211)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

TODO(bassosimone): investigate this
</blockquote></details>

I would not worry about this until we profile the running system.

---

*[protocol/protocol.go, line 109 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/29#-LEqF5pL-rOUEG3yNi0c:-LEqF5pL-rOUEG3yNi0d:b-co864j) ([raw file](https://github.com/m-lab/ndt-server-go/blob/37e6a43eb54e1d7209858173d64afcd38f185c2b/protocol/protocol.go#L109)):*
> ```Go
> // RecvBinaryMessage reads a binary NDT message from |brdr|. Returns the message
> // and/or the error that occurred while reading such message.
> func RecvBinaryMessage(brdr *bufio.Reader) (Message, error) {
> ```

This should probably be an io.Reader, instead of buffered reader.  I don't think you are using any of the bufio features.

---

*[protocol/protocol.go, line 121 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/29#-LEqEdwK84ah6u8m-Diz:-LEqEdwK84ah6u8m-Dj-:b8ly6s4) ([raw file](https://github.com/m-lab/ndt-server-go/blob/37e6a43eb54e1d7209858173d64afcd38f185c2b/protocol/protocol.go#L121)):*
> ```Go
> 	log.Printf("ndt: message length: %d", hdr.Length)
> 	content := make([]byte, hdr.Length)
> 	// TODO(bassosimone): discuss with @gfr10598 whether here it makes
> ```

IIUC, it shouldn't matter, so OK to use this, or use a simple reader.  I'd say since you are using binary.Read above, might as well use it here too.

---


*Comments from [Reviewable](https://reviewable.io/reviews/m-lab/ndt-server-go/29)*

**Review by bassosimone — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/m-lab/ndt-server-go/29)* status: 0 of 1 LGTMs obtained (waiting on @bassosimone)

___
*[control/control.go, line 117 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/29#-LEqENUc8zXiy4yRmi9N:-LIjrQWA8ZHD3XaEwIUH:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/37e6a43eb54e1d7209858173d64afcd38f185c2b/control/control.go#L117)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This is no longer correct.  Want to update to refer to tcp_info?
</blockquote></details>

Done.
___
*[protocol/protocol.go, line 207 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/29#-LEFSDBbM8z5iCr2MC52-r1-207:-LIjx01T0-fD6Oeqt08M:b-80tlqx) ([raw file](https://github.com/m-lab/ndt-server-go/blob/a92591bbe0b7b4e9bb357317b7aef3aeffb9d0be/protocol/protocol.go#L207)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I don't think you should be panicing.  Let's discuss.
</blockquote></details>

Sure!
___
*[protocol/protocol.go, line 211 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/29#-LEFSDBbM8z5iCr2MC52-r1-211:-LIjxEeO5MZcV1FFwjjA:b32w0x) ([raw file](https://github.com/m-lab/ndt-server-go/blob/a92591bbe0b7b4e9bb357317b7aef3aeffb9d0be/protocol/protocol.go#L211)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

I would not worry about this until we profile the running system.
</blockquote></details>

OK
___
*[protocol/protocol.go, line 109 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/29#-LEqF5pL-rOUEG3yNi0c:-LIjv9hrEG0c6kNCpf_o:b-iyosvq) ([raw file](https://github.com/m-lab/ndt-server-go/blob/37e6a43eb54e1d7209858173d64afcd38f185c2b/protocol/protocol.go#L109)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This should probably be an io.Reader, instead of buffered reader.  I don't think you are using any of the bufio features.
</blockquote></details>

Let's discuss this. I initially thought the a buffered reader was good to protect against the case where back to back messages are sent. But now I am not sure about this anymore and I probably need some clarifications. Is it correct to say that:

```Go
var hdr header
binary.Read(brdr, binary.BigEndian, &hdr)
```

is equivalent to

```C++
class Header header;
::recv(fd, &header, sizeof (header));
```

That is, is it correct to say that in both cases, on success, just the amount of bytes required to extract a binary message are required? If that is correct, then I think I had misunderstood how Golang reads from sockets when I implemented botticelli. In such case, some of the code I wrote there and ported in here would strike me as extra complexity. Since now I have a sense that we always read exactly the amount that we need from the socket, hence there is no need for any buffered I/O at all.
___
*[protocol/protocol.go, line 121 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/29#-LEqEdwK84ah6u8m-Diz:-LIjwN7J0uhnS8p8u0Dv:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/37e6a43eb54e1d7209858173d64afcd38f185c2b/protocol/protocol.go#L121)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

IIUC, it shouldn't matter, so OK to use this, or use a simple reader.  I'd say since you are using binary.Read above, might as well use it here too.
</blockquote></details>

Done.

**Comment by gfr10598** [comment]
Obsolete.

**Pull request closed without merging**